### ISSUE TITLE
Component Generics

### ISSUE THREAD

**Issue by naturalethic** [DioxusLabs/dioxus#321]
Allow for generics on components.  Consider a `Form` abstraction.  Wanting to reuse the design and basic mechanics of that form, but provide specialized data, I could provide a data generic such as `Form<Registration>` which struct may implement a trait for validation, submission, etc.

```
div {
   Form<Registration> {
    data: registration_data,  /// type "Registration"
  }
}
```

Just a simple example, but I think generics in the context of components would be pretty useful generally.

**Comment by rMazeiks** [comment]
I hacked up a simple example, and generic components seem to work... You don't need that <> syntax in the rsx though.

```rust
#![allow(non_snake_case)]

use std::fmt::Display;
use dioxus::prelude::*;

fn main() {
    dioxus::desktop::launch(App);
}

fn App(cx: Scope) -> Element {
    let integer = 5;
    let string = "string".to_string();

    cx.render(rsx! (
        div {
            DisplayComponent {
                value: integer
            }
            DisplayComponent {
                value: string
            }
        }
    ))
}

#[derive(PartialEq, Props)]
pub struct DisplayProps<T: Display> {
    value: T
}

pub fn DisplayComponent<T:Display>(cx: Scope<DisplayProps<T>>) -> Element {
    let formatted = format!("Currently on display: {}", cx.props.value);

    cx.render(rsx!(
        div {"{formatted}"}
    ))
}
```

**Comment by naturalethic** [comment]
Oh interesting, thanks!

**Comment by MuhannadAlrusayni** [comment]
I'm trying to make fully typed components for input/forms and I'm blocked by the fact Dioxus doesn't support generic components (at less full support).

Unfortunately @rMazeiks solution doesn't work for me, because this solution only work when you pass **value** of the generic type within the props values, and in my case I want to pass the **type** it self not value of it.

Here is the design I'm trying to get it work:
```rust
#[derive(..)]
pub Struct SignIn {
  name: Username,
  password: Password,
}

rsx!(
  // Here I pass the struct type that contains all fields the we
  // will get in `on_valid` event handler, and all it's field was validated
  // using `FromStr`.
  Form<SignIn> {
    // in this handler we can do what ever with the sign_in value,
    // e.g. send it to API endpoint
    on_valid: move |sign_in: SignIn| todo!(),
    
    // note here I pass the field type to Input so it can use `FromStr`
    // for validation
    Input<Username> { name: "name" }
    Input<Password> { name: "password" }
  }
)
```

**Comment by rMazeiks** [comment]
Oooh, interesting project! Indeed, I believe what you're trying to achieve is not currently possible. However, it also looks like your API would be a bit difficult to use. In particular, here are some concerns:

- What if multiple input fields have the same type? For example, `password` and `verify_password`... how is the library supposed to "match" Inputs to fields?
- The API is not as type-safe as you might think – if you render, for example, a `Input<String>` inside that form, it will compile even though there is no such field. This is because components' children are not checked at compile time.
- In your design, it looks like all of the state lives inside the `Form<SignIn>` component, so you have no way of accessing it from outside. What if you want to programmatically change it? (e.g. autofill, username suggestions, "reset" button, etc.).

May I suggest an alternative API:

Create a macro that implements the field validation logic in a hook. You'd have to be a bit more explicit with how the state is passed around, but that might be a good thing – it's more clear where the data flows.

```rust
// this macro automatically creates:
// - the SignIn::Form struct
// - the SignIn::use_form hook
#[auto_form]
struct SignIn {
  name: Username,
  password: Password,
}

let form_data = SignIn::use_form();

// these variables (name, password) would be similar to the output of the `use_state` hook
// that is, you can read their current value and set it.
// they might implement some validation too
let SignIn::Form { name, password } = form_data.state()

// you could access the final `SignIn` object using something like `form_data.build()`

let validation = form_data.validation_message();

rsx!(
  form {
    on_submit: move || todo!(),
    
    Input { name: "name", value: name }
    Input { name: "password", value: password }
    "{validation}"
  }
)
```

Now I admit, there is less "automagic" going on. But it has some advantages:

- It's clear how input fields map to struct fields
- You can't accidentally add an input for a nonexistent field, or forget about a field – it's all typechecked.
- You have full access to the current state, so you can update it, or even perform additional validation in the component

anyway, that's just an idea. it's possible i haven't thought of something, i didn't try any of this.

**Comment by MuhannadAlrusayni** [comment]
The design I have shown was just an example, talking about the design of Form/Input is kinda out of scope of this issue.

The thing we need to talk about is that we need Generic Component. and I'm interested into make PR for this but I need a mentor :D since this is my 1st PR.

**Comment by jkelleyrtp** [comment]
> The design I have shown was just an example, talking about the design of Form/Input is kinda out of scope of this issue.
> 
> The thing we need to talk about is that we need Generic Component. and I'm interested into make PR for this but I need a mentor :D since this is my 1st PR.

If you want to hop in the internals channel on discord I can show you where you would need to add parsing for generics in the rsx macro crate. Generics would be very easy add, I think.

**Comment by oskgo** [comment]
I have a different use case for this. I have an app that interacts with a third party application and I want to do dependency injection to simplify testing. I really want to do this using the type system similar to how the Allocator API works, but this is currently impossible because generics don't work.

**Comment by jkelleyrtp** [comment]
Solved by #385 . Woohoo!
